File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.java
class Solution {
    public String shortestBeautifulSubstring(String s, int k) {
        int n = s.length();
        int lft = 0 , ones = 0;

        String ans = "";

        for(int rght = 0 ; rght < n ; rght++){

            if(s.charAt(rght) == '1')
                ones++;

            while(ones > k){
                if(s.charAt(lft) == '1')
                    ones--;
                lft++;
            }

            if(ones == k){
                while(s.charAt(lft) == '0')
                    lft++;

                String curr = s.substring(lft , rght + 1);

                if(ans.equals("") || ans.length() > curr.length() || (ans.length() == curr.length() && curr.compareTo(ans) < 0))
                    ans = curr;
            }
        }

        return ans;
    }
}

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        n = len(s)

        ones = 0
        lft = 0

        ans = ""

        for rght in range(0 , n):
            if s[rght] == '1':
                ones += 1

            while ones > k :
                if s[lft] == '1':
                    ones -= 1

                lft += 1

            if ones == k:
                while s[lft] == '0':
                    lft += 1

                curr = s[lft : rght + 1]

                if len(ans) > len(curr) or ans == "" or (ans > curr and len(ans) == len(curr)):
                    ans = curr

        return ans

                